fendermint/vm/topdown/src/convert.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! Handles the type conversion to ethers contract types
+
+use crate::IPCParentFinality;
+use anyhow::anyhow;
+use ethers::abi::{Function, Token};
+use ethers::types::Bytes;
+use ethers::types::U256;
+use fendermint_vm_ipc_actors::{gateway_getter_facet, gateway_router_facet};
+use fvm_shared::address::{Address, Payload};
+use ipc_agent_sdk::message::ipc::ValidatorSet;
+use std::str::FromStr;
+
+const COMMIT_PARENT_FINALITY_FUNC_NAME: &str = "commitParentFinality";
+const GET_LATEST_PARENT_FINALITY_FUNC_NAME: &str = "getLatestParentFinality";
+
+impl TryFrom<IPCParentFinality> for gateway_router_facet::ParentFinality {
+    type Error = anyhow::Error;
+
+    fn try_from(value: IPCParentFinality) -> Result<Self, Self::Error> {
+        if value.block_hash.len() != 32 {
+            return Err(anyhow!("invalid block hash length, expecting 32"));
+        }
+
+        let mut block_hash = [0u8; 32];
+        block_hash.copy_from_slice(&value.block_hash[0..32]);
+
+        Ok(Self {
+            height: U256::from(value.height),
+            block_hash,
+        })
+    }
+}
+
+impl From<gateway_getter_facet::ParentFinality> for IPCParentFinality {
+    fn from(value: gateway_getter_facet::ParentFinality) -> Self {
+        IPCParentFinality {
+            height: value.height.as_u64(),
+            block_hash: value.block_hash.to_vec(),
+        }
+    }
+}
+
+/// Converts a Rust type FVM address into its underlying payload
+/// so it can be represented internally in a Solidity contract.
+fn addr_payload_to_bytes(payload: Payload) -> anyhow::Result<Bytes> {
+    let r = match payload {
+        Payload::Secp256k1(v) => ethers::types::Bytes::from(v),
+        Payload::Delegated(d) => {
+            let addr = d.subaddress();
+            let b = ethers::abi::encode(&[Token::Tuple(vec![
+                Token::Uint(U256::from(d.namespace())),
+                Token::Uint(U256::from(addr.len())),
+                Token::Bytes(addr.to_vec()),
+            ])]);
+            ethers::types::Bytes::from(b)
+        }
```

Comment:
I don't know how we can handle delegated addresses. At the end of the day we need to give Tendermint a public key and this is the only place I will learn it from.

Reply:
From the discussion on Slack, we can try creating our own address namespace.